src/change.c

```diff
@@ -149,8 +149,91 @@ changed_internal(void)
 }
 
 #ifdef FEAT_EVAL
+// Set when listener callbacks are being invoked.
+static int recursive = FALSE;
+
 static long next_listener_id = 0;
 
+// A flag that is set when any buffer listener housekeeping is required.
+// Currently the only condition is when a listener is marked for removal.
+static bool houskeeping_required;
+
+/*
+ * Remove a given listener_T entry from its containing list.
+ */
+    static void
+remove_listener_from_list(
+    listener_T **list,
+    listener_T *lnr,
+    listener_T *prev)
+{
+    if (prev != NULL)
+	prev->lr_next = lnr->lr_next;
+    else
+	*list = lnr->lr_next;
+    free_callback(&lnr->lr_callback);
+    vim_free(lnr);
+}
+
+/*
+ * Clean up a buffer change listener list.
+ *
+ * If "all" is TRUE then all entries are removed. Otherwise only those with an ID
+ * of zero are removed. If "buf" is non-NULL then the buffer's recorded changes
+ * will be discarded in the event that all listeners were removed.
+ *
+ */
+    static void
+clean_listener_list(buf_T *buf, listener_T **list, bool all)
+{
+    listener_T	*prev;
+    listener_T	*lnr;
+    listener_T	*next;
+
+    prev = NULL;
+    for (lnr = *list; lnr != NULL; lnr = next)
+    {
+	next = lnr->lr_next;
+	if (all || lnr->lr_id == 0)
+	{
```

Comment:
style: you can drop braces around single statements

Reply:
Agreed.